### PR TITLE
Fix segfault in nested tuple to tuple assignments

### LINKED ISSUE #13901: Null dereference in `checkDoubleStorageAssignment()` when type-checking some forms of tuple assignment expressions
```
contract C{bool x;fallback(){(((),x)&x)=((),0);}}
```

```
solc test.sol
```

throws due to null deref from here

https://github.com/ethereum/solidity/blob/206e7cf5db48b13085b943436764c41f169c6c3b/libsolidity/analysis/TypeChecker.cpp#L119

### PR DESCRIPTION
**Pull request by nikola-matic**
Fixes https://github.com/ethereum/solidity/issues/13901.

### CODE REVIEW

File: libsolidity/analysis/StaticAnalyzer.cpp [hunk 1]
```diff
@@ -22,16 +22,30 @@
...
+
+#include <range/v3/view/enumerate.hpp>
+
 #include <memory>
 
 using namespace solidity;
 using namespace solidity::langutil;
 using namespace solidity::frontend;
 
+namespace
```
Comment (nikola-matic) [inline libsolidity/analysis/StaticAnalyzer.cpp]: This is a copy of the same function from [`TypeChecker`](https://github.com/ethereum/solidity/blob/develop/libsolidity/analysis/TypeChecker.cpp#L83-L93) which I'd like to move into a common utils file. Would `ASTUtils` be appropriate for this? I'd do this in another PR of course.

File: libsolidity/analysis/StaticAnalyzer.cpp [hunk 2]
```diff
@@ -348,3 +374,100 @@ bool StaticAnalyzer::visit(FunctionCall const& _functionCall)
...
+{
+	size_t storageToStorageCopies = 0;
+	size_t toStorageCopies = 0;
+	size_t storageByteArrayPushes = 0;
+	size_t storageByteAccesses = 0;
+	auto count = [&](TupleExpression const& _lhs, TupleType const& _rhs, auto _recurse) -> void {
+		TupleType const& lhsType = dynamic_cast<TupleType const&>(*type(_lhs));
+		TupleExpression const* lhsResolved = dynamic_cast<TupleExpression const*>(resolveOuterUnaryTuples(&_lhs));
+
+		solAssert(lhsResolved && lhsResolved->components().size() == lhsType.components().size());
```
Comment (nikola-matic) [inline libsolidity/analysis/StaticAnalyzer.cpp:377]: This assert has now been strengthened to make sure that `lhsResolved` is not a nullptr.

File: libsolidity/analysis/StaticAnalyzer.cpp (hunk 1 above)
Comment (nikola-matic) [reply libsolidity/analysis/StaticAnalyzer.cpp]: I did it anyway, it's better this way https://github.com/ethereum/solidity/pull/15148.

File: libsolidity/analysis/StaticAnalyzer.cpp [hunk 3]
```diff
@@ -22,16 +22,30 @@
...
 using namespace solidity::langutil;
 using namespace solidity::frontend;
 
+namespace
+{
+
+Type const* type(Expression const& _expression)
+{
+	solAssert(!!_expression.annotation().type, "Type requested but not present.");
+	return _expression.annotation().type;
```
Comment (clonker) [inline libsolidity/analysis/StaticAnalyzer.cpp]: you could change the return value to a `Type const&`  if the expectation is that it exists and it is also asserted for

File: libsolidity/analysis/StaticAnalyzer.cpp [hunk 4]
```diff
@@ -348,3 +374,100 @@ bool StaticAnalyzer::visit(FunctionCall const& _functionCall)
...
+		{
+			solAssert(m_errorReporter.hasErrors(), "");
+			return;
+		}
+
+		for (auto&& [index, componentType]: lhsType.components() | ranges::views::enumerate)
+		{
+			if (ReferenceType const* ref = dynamic_cast<ReferenceType const*>(componentType))
+			{
+				if (ref && ref->dataStoredIn(DataLocation::Storage) && !ref->isPointer())
```
Comment (clonker) [inline libsolidity/analysis/StaticAnalyzer.cpp]:
```suggestion
				if (ref->dataStoredIn(DataLocation::Storage) && !ref->isPointer())
```
`ref` must exist here as

File: libsolidity/analysis/StaticAnalyzer.cpp [hunk 5]
```diff
@@ -348,3 +374,100 @@ bool StaticAnalyzer::visit(FunctionCall const& _functionCall)
...
+				{
+					toStorageCopies++;
+					if (_rhs.components()[index]->dataStoredIn(DataLocation::Storage))
+						storageToStorageCopies++;
+				}
+			}
+			else if (FixedBytesType const* bytesType = dynamic_cast<FixedBytesType const*>(componentType))
+			{
+				if (bytesType && bytesType->numBytes() == 1)
```
Comment (clonker) [inline libsolidity/analysis/StaticAnalyzer.cpp]:
```suggestion
				if (bytesType->numBytes() == 1)
```

File: libsolidity/analysis/StaticAnalyzer.cpp (hunk 3 above)
Comment (clonker) [reply libsolidity/analysis/StaticAnalyzer.cpp]: matter of a different PR

File: test/libsolidity/syntaxTests/tupleAssignments/nested_tuple_to_tuple_assignment.sol [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    function foo() external {
+        (((),x)&x) = ((), true);
+        (((),x)+x) = ((), true);
+        (((),x)-x) = ((), true);
+        (((),x)/x) = ((), true);
+        (((),x)*x) = ((), true);
+        (((),x)|x) = ((), true);
+    }
+}
+//----
```
Comment (clonker) [inline test/libsolidity/syntaxTests/tupleAssignments/nested_tuple_to_tuple_assignment.sol]:
```suggestion
```

**Review by clonker — APPROVED** [review]

**Pull request merged**